### PR TITLE
Add the Gatekeeper constraint status sync controller

### PR DESCRIPTION
**Pull request by mprahl**
This controller is responsible for relaying Gatekeeper constraint audit results as policy status events for constraints associated with a policy.

The controller will only start if Gatekeeper is installed. If Gatekeeper's installation status changes while the controller manager is running, the health endpoint will return as unhealthy to cause Kubernetes to restart the pod to then enable/disable this new controller.

Relates:
https://issues.redhat.com/browse/ACM-3328

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -363,6 +371,89 @@ func getManager(
...
+		"governance-policy-framework-addon", tool.Options.HubConfigFilePathName,
+	)
+	if err != nil {
+		log.Error(err, "unable to setup a configChecker")
+		os.Exit(1)
+	}
+
+	healthCheck := configChecker.Check
+
+	// Gatekeeper is not supported on such old versions of Kubernetes, so just always disable it in this case.
```
Comment (JustinKuli) [inline main.go:393]: I think this comment is confusing, since we're not conditionally *disabling* it, we're conditionally *enabling* it.
```suggestion
	// Only enable the GatekeeperConstraintReconciler if the Kubernetes version is v1.16 or above
```

File: controllers/gatekeepersync/health.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+) {
+	fieldSelector := fmt.Sprintf("metadata.name=constrainttemplates.%s", utils.GvkConstraintTemplate.Group)
+	timeout := int64(30)
+	crdGVR := schema.GroupVersionResource{
+		Group:    "apiextensions.k8s.io",
+		Version:  "v1",
+		Resource: "customresourcedefinitions",
+	}
+
+	listResult, err := dynamicClient.Resource(crdGVR).List(
```
Comment (JustinKuli) [inline controllers/gatekeepersync/health.go:39]: If this is the first thing done by the dynamic client (and I think it might be) then I think it could take a long time to do its API discovery... you might already know more about that than me, so correct me if we don't need to be concerned with that here. But if this `List` times out (because of that issue), then the error would bubble up to `main`, and we'd effectively restart the container. But restarting the container will just cause it to have the same problem again, unless I'm missing something.

File: controllers/gatekeepersync/health.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+		if lastHealthErr != nil {
+			return lastHealthErr
+		}
+
+		select {
+		case <-w.Done():
+			select {
+			case <-ctx.Done():
+				// Stop the retry watcher if the parent context is canceled.
+				w.Stop()
```
Comment (JustinKuli) [inline controllers/gatekeepersync/health.go:72]: If the parent context was canceled, should we return a non-nil error? Otherwise it feels like there *could* be some weird situation where it never appears unhealthy, because it keeps hitting this spot?

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+			r.lastSentMessages.Range(func(key, value any) bool {
+				keyTyped := key.(policyKindName)
+				if keyTyped.Policy == request.Name {
+					r.lastSentMessages.Delete(keyTyped)
+				}
+
+				return true
+			})
+
+			err := r.ConstraintsWatcher.RemoveWatcher(policyObjID)
```
Comment (JustinKuli) [inline controllers/gatekeepersync/gatekeeper_constraint_sync.go:132]: I believe this only returns a non-nil error if the policyObjID is invalid... it shouldn't be invalid, so it should always be nil, but if it *was* somehow invalid then the error here would cause the policy to be re-queued forever.

**Review by JustinKuli — CHANGES_REQUESTED** [review]
This is only a partial review, but I had some early questions.

**Review by dhaiducek — COMMENTED** [review]
I gave this a first pass, and I don't think I saw anything glaring. I'll look at it again first thing tomorrow when my vviissiioonn isn't blurring. 😉

File: Makefile [hunk 5]
```diff
@@ -283,6 +283,10 @@ install-resources:
 	-kubectl create ns $(MANAGED_CLUSTER_NAME) --kubeconfig=$(HUB_CONFIG)
 	@echo creating namespace on managed
 	-kubectl create ns $(MANAGED_CLUSTER_NAME) --kubeconfig=$(MANAGED_CONFIG)
+	@if [ "$(KIND_VERSION)" != "minimum" ]; then \
+		echo installing Gatekeeper on the managed cluster; \
+		kubectl apply -f https://raw.githubusercontent.com/open-policy-agent/gatekeeper/v3.11.0/deploy/gatekeeper.yaml  --kubeconfig=$(MANAGED_CONFIG); \
+	fi
```
[the remark this answers is not in the document]
Comment (dhaiducek) [reply Makefile:289]: It follows from technical debt in my PR:
> Handling for older Kubernetes/older Gatekeeper (right now, I have a skip-minimum tag that skips the test entirely); Gatekeeper v3.4 added a PodDisruptionBudget that isn't available until Kubernetes v1.21

https://github.com/open-cluster-management-io/governance-policy-framework-addon/pull/33#issuecomment-1450652169

File: test/e2e/case17_gatekeeper_sync_test.go [hunk 6]
```diff
@@ -42,31 +51,102 @@ var _ = Describe("Test Gatekeeper ConstraintTemplate sync", Ordered, Label("skip
...
+			var available bool
+			for _, deployment := range deployments.Items {
+				for _, condition := range deployment.Status.Conditions {
+					if condition.Reason == "MinimumReplicasAvailable" {
+						available = condition.Status == "True"
+					}
+				}
+				g.Expect(available).To(BeTrue())
+			}
+		}, defaultTimeoutSeconds, 1).Should(Succeed())
```
[the remark this answers is not in the document]
Comment (mprahl) [reply test/e2e/case17_gatekeeper_sync_test.go:84]: Good catch.

File: controllers/utils/events.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	policyv1 "open-cluster-management.io/governance-policy-propagator/api/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type EventSender interface {
+	GetInstanceName() string
+	GetClusterNamespace() string
+	GetClientSet() *kubernetes.Clientset
+	GetControllerName() string
+}
```
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/utils/events.go:22]: I felt like it was clunky as well. I'm going to embed a struct with a `SendEvent` method that I think will be cleaner.

File: controllers/utils/events.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+// SendComplianceEvent will send a policy template status message update synchronously as opposed to EventRecorder
+// sending events in the background asynchronously.
+func SendComplianceEvent(
+	ctx context.Context,
+	e EventSender,
+	instance client.Object,
+	owner metav1.OwnerReference,
```
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/utils/events.go:30]: I think it's good enough for the moment. If you come up with an alternative, please let me know.

File: main.go (hunk 1 above)
Comment (mprahl) [reply main.go:393]: I think I'll remove this semver check and just assume if Gatekeeper is installed, it is working.

File: controllers/gatekeepersync/health.go (hunk 2 above)
Comment (mprahl) [reply controllers/gatekeepersync/health.go:39]: The dynamic client doesn't perform a discovery from what I can tell in the code. It just uses the GVR to construct the Kubernetes API endpoint URL.

File: controllers/gatekeepersync/health.go (hunk 3 above)
Comment (mprahl) [reply controllers/gatekeepersync/health.go:72]: Yes, good point.

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+		APIVersion: refreshedPolicy.APIVersion,
+		Kind:       refreshedPolicy.Kind,
+		Name:       refreshedPolicy.Name,
+		UID:        refreshedPolicy.UID,
+	}
+
+	if len(refreshedPolicy.Status.Details) < templateIndex+1 ||
+		len(refreshedPolicy.Status.Details[templateIndex].History) == 0 ||
+		refreshedPolicy.Status.Details[templateIndex].History[0].Message != fmt.Sprintf("%s; %s", compliance, msg) {
+		err = utils.SendComplianceEvent(ctx, r, constraint, owner, msg, compliance)
```
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/gatekeepersync/gatekeeper_constraint_sync.go:366]: I'll rework this so that if it attempts to send the same status message and it's already in the Policy status, the entry in the lastSentMessages is deleted.

File: test/e2e/case17_gatekeeper_sync_test.go [hunk 10]
```diff
@@ -161,6 +245,205 @@ var _ = Describe("Test Gatekeeper ConstraintTemplate sync", Ordered, Label("skip
...
+
+			history := managedPolicy.Status.Details[1].History
+			g.Expect(len(history)).ToNot(Equal(0))
+			expectedMsg := `NonCompliant; The Gatekeeper validating webhook is disabled but the constraint's ` +
+				`spec.enforcementAction is deny. deny - All configmaps must have a 'my-gk-test' label (on ConfigMap ` +
+				`case17-gk-test/case17-test); deny - All configmaps must have a 'my-gk-test' label (on ConfigMap ` +
+				`case17-gk-test/case17-test2)`
+			g.Expect(history[0].Message).To(
+				Equal(expectedMsg),
+				fmt.Sprintf("Got %s but expected %s", history[0].Message, expectedMsg),
```
[the remark this answers is not in the document]
Comment (mprahl) [reply test/e2e/case17_gatekeeper_sync_test.go:421]: It has a similar message but it truncates it so much that it's hard to debug.

File: controllers/gatekeepersync/policy_predicate.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// policyPredicates filters out policies without Gatekeeper constraints and policy updates without the generation
+// changing.
+func policyPredicates() predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			// nolint: forcetypeassert
```
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/gatekeepersync/policy_predicate.go:16]: It looks like it wasn't needed. I guess it was preemptive on my part.  :laughing:

[the remark this answers is not in the document]
Comment (JustinKuli) [reply controllers/gatekeepersync/policy_predicate.go]: Maybe we need to re-enable the linter somewhere? Because I would rather not have panics 😄

**Review by JustinKuli — DISMISSED** [review]
/hold

Looks good to me! Thanks for those updates. Leaving it up to @dhaiducek to double-check :)

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go [hunk 12]
```diff
@@ -0,0 +1,426 @@
...
+)
+
+var log = logf.Log.WithName(ControllerName)
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *GatekeeperConstraintReconciler) SetupWithManager(mgr ctrl.Manager, constraintEvents *source.Channel) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&policyv1.Policy{}).
+		WithEventFilter(policyPredicates()).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 5}).
```
Comment (dhaiducek) [inline controllers/gatekeepersync/gatekeeper_constraint_sync.go:47]: There isn't concurrency in the other controllers that I can see. Concurrency seems good, but I'm wondering about the effect here.

File: test/e2e/case17_gatekeeper_sync_test.go [hunk 13]
```diff
@@ -125,6 +208,205 @@ var _ = Describe("Test Gatekeeper ConstraintTemplate sync", Ordered, Label("skip
...
+			g.Expect(managedPolicy.Status.Details).To(HaveLen(2))
+			g.Expect(managedPolicy.Status.Details[1].TemplateMeta.GetName()).To(Equal(gkConstraintName))
+			history := managedPolicy.Status.Details[1].History
+			g.Expect(len(history)).ToNot(Equal(0))
+			expectedMsg := "Compliant; The constraint has no violations"
+			g.Expect(history[0].Message).To(
+				Equal(expectedMsg),
+				fmt.Sprintf("Got %s but expected %s", history[0].Message, expectedMsg),
+			)
+		}, "120s", 1).Should(Succeed())
```
Comment (dhaiducek) [inline test/e2e/case17_gatekeeper_sync_test.go:234]: Nit: Following other conventions here. Is there a reason to hardcode this value (as well as the other instances in this suite)?
```suggestion
		}, defaultTimeoutSeconds*4, 1).Should(Succeed())
```

**Review by dhaiducek — APPROVED** [review]
LGTM!
I have a musing about concurrency here where it doesn't seem to be elsewhere and using `defaultTimeoutSeconds` consistently, but they're not anything that should block the PR.

File: controllers/gatekeepersync/gatekeeper_constraint_sync.go (hunk 12 above)
Comment (mprahl) [reply controllers/gatekeepersync/gatekeeper_constraint_sync.go:47]: @dhaiducek you're right. I added concurrency here because by default, the Gatekeeper audits run every minute so if the user has a lot of constraints, I didn't want there to be a delay in getting results.

The way the concurrency works in controller-runtime is that there is only one goroutine processing a particular reconcile request at a time. For example, since we are reconciling `Policy` objects, if there are three reconciles requests for `Policy` named `a` and one reconcile request for `Policy` named `b`, there would just be one goroutine processing `Policy` `a` and one goroutine processing `Policy` `b`.

In other words, it's concurrent across reconciles for different `Policy` objects but reconciles for the same `Policy` will not run concurrently.

File: test/e2e/case17_gatekeeper_sync_test.go (hunk 13 above)
Comment (mprahl) [reply test/e2e/case17_gatekeeper_sync_test.go]: I changed this to use a different constant.
